File: src/profiles.py
from pathlib import Path


def load_profiles(controls_dir: Path, user_dir: Path | None) -> dict[str, str]:
    """Load zero, light, and any user-supplied profiles.

    Returns a dict mapping profile name to system prompt text.
    Zero profile is always an empty string.
    Light profile is the contents of controls_dir/light.txt.
    User profiles are loaded from user_dir (*.txt and *.md files).
    Control name collisions (zero.txt, light.txt) are skipped with a warning.
    """
    _CONTROL_NAMES = {"zero", "light"}

    profiles: dict = {
        "zero": "",
        "light": (controls_dir / "light.txt").read_text().strip(),
    }
    if user_dir and user_dir.exists():
        for f in sorted(user_dir.glob("*.txt")) + sorted(user_dir.glob("*.md")):
            if f.stem in _CONTROL_NAMES:
                print(f"WARNING: skipping user profile '{f.name}' — collides with control baseline")
                continue
            profiles[f.stem] = f.read_text().strip()
    return profiles

File: internal/profiles/profiles.go
package profiles

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var controlNames = map[string]bool{"zero": true, "light": true}

// LoadProfiles loads the zero, light, and any user-supplied profiles.
// controlsDir must contain light.txt.
// userDir may be empty string or non-existent (both are silently skipped).
// Returns map of profile name -> system prompt text.
func LoadProfiles(controlsDir, userDir string) (map[string]string, error) {
	lightPath := filepath.Join(controlsDir, "light.txt")
	lightBytes, err := os.ReadFile(lightPath)
	if err != nil {
		return nil, fmt.Errorf("reading light profile: %w", err)
	}

	profiles := map[string]string{
		"zero":  "",
		"light": strings.TrimSpace(string(lightBytes)),
	}

	if userDir == "" {
		return profiles, nil
	}
	info, err := os.Stat(userDir)
	if err != nil || !info.IsDir() {
		return profiles, nil
	}

	// Collect *.txt and *.md files, sort them for determinism.
	var files []string
	for _, pat := range []string{"*.txt", "*.md"} {
		matches, _ := filepath.Glob(filepath.Join(userDir, pat))
		files = append(files, matches...)
	}
	sort.Strings(files)

	for _, path := range files {
		stem := strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
		if controlNames[stem] {
			fmt.Printf("WARNING: skipping user profile '%s' — collides with control baseline\n", filepath.Base(path))
			continue
		}
		content, err := os.ReadFile(path)
		if err != nil {
			return nil, fmt.Errorf("reading user profile %s: %w", path, err)
		}
		profiles[stem] = strings.TrimSpace(string(content))
	}
	return profiles, nil
}
